### PR TITLE
Support Rails 7

### PR DESCRIPTION
**Pull request by skateman**
Taking it over from https://github.com/zendesk/stronger_parameters/pull/93

### CODE REVIEW

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb [hunk 1]
```diff
@@ -28,6 +28,8 @@ def show
   def get(action, options)
     if Rails::VERSION::MAJOR < 5
       super(action, options.fetch(:params).merge(format: options[:format] || 'html'))
```
Comment (grosser) [inline test/stronger_parameters/controller_support/permitted_parameters_test.rb:30]: can delete all the `< 5` code

File: test/controller_test.rb [hunk 2]
```diff
@@ -24,6 +24,8 @@ def create
   def post(action, options = {})
     if Rails::VERSION::MAJOR < 5
       super(action, options.fetch(:params))
+    elsif Rails::VERSION::MAJOR >= 7
+      super(action, **options)
```
Comment (grosser) [inline test/controller_test.rb:28]: kinda the same as the else so try removing it 🤞

File: test/controller_test.rb [hunk 3]
```diff
@@ -21,13 +21,15 @@ def create
...
 
 describe BooksController do
-  def post(action, options = {})
-    if Rails::VERSION::MAJOR < 5
-      super(action, options.fetch(:params))
-    else
-      super
-    end
-  end
+  # def post(action, options = {})
```
Comment (grosser) [inline test/controller_test.rb:24]: delete

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb (hunk 1 above)
Comment (grosser) [reply test/stronger_parameters/controller_support/permitted_parameters_test.rb:30]: ^

File: test/test_helper.rb [hunk 4]
```diff
@@ -6,10 +6,14 @@
...
 
+require 'rails'
 require 'maxitest/global_must'
 require 'maxitest/autorun'
-require 'mocha/setup'
-require 'rails'
+if Rails::VERSION::MAJOR >= 7
+  require 'mocha/minitest'
+else
+  require 'mocha/setup'
```
Comment (grosser) [inline test/test_helper.rb:15]: can we update mocha in the lower rails versions to get rid of this ?

**Comment by skateman** [comment]
@grosser I'm trying to do it in multiple steps as I'm struggling with old ruby versions in my env

File: test/stronger_parameters/constraints/hash_constraint_test.rb [hunk 5]
```diff
@@ -1,7 +1,7 @@
 # frozen_string_literal: true
 require_relative '../../test_helper'
 
-SingleCov.covered! uncovered: (ActiveSupport::VERSION::MAJOR < 5 ? 1 : 0)
+SingleCov.covered! uncovered: 0
```
Comment (grosser) [inline test/stronger_parameters/constraints/hash_constraint_test.rb:4]: FYI
```suggestion
SingleCov.covered!
```

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb [hunk 6]
```diff
@@ -2,7 +2,7 @@
 require_relative '../../test_helper'
 require 'stronger_parameters/controller_support/permitted_parameters'
 
-SingleCov.covered! uncovered: (ActiveSupport::VERSION::MAJOR < 5 ? 3 : 2) # uncovered branches for rails version check
+SingleCov.covered! uncovered: 2
```
Comment (grosser) [inline test/stronger_parameters/controller_support/permitted_parameters_test.rb:5]: ideally remove `uncovered: 2` here and leave magic `# uncovered` comments in the lines that are uncovered

**Comment by skateman** [comment]
@grosser I have no idea how I could recreate the GH actions environment in order to have the right lockfiles :disappointed: can you help me please?

**Review by grosser — APPROVED** [review]
wow so much cleanup :D

**Comment by grosser** [comment]
`BUNDLE_GEMFILE=gemfiles/foo.gemfile bundle exec rake` should do it
or is there any error you run into ?

**Comment by skateman** [comment]
@grosser yeah, that's what I used and you see that the CI doesn't like the lockfiles it's generating ...

**Comment by grosser** [comment]
if nothing works then 
```
    env:
      BUNDLE_DEPLOYMENT: false
```
🙈

**Comment by grosser** [comment]
I can give it a try tomorrow too ... usually it just worked 🤷

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb (hunk 6 above)
Comment (skateman) [reply test/stronger_parameters/controller_support/permitted_parameters_test.rb:5]: I don't know which lines are uncovered, so I'm keeping it as it is for now and will play with it after tests pass.

**Comment by skateman** [comment]
Seems like deployment can't be disabled if the lockfile is present :sob:

**Comment by grosser** [comment]
https://github.com/ruby/setup-ruby/issues/292 says it does not work 😞

**Comment by grosser** [comment]
I'll poke it tomorrow 🤞

**Comment by grosser** [comment]
diff after a clean bundle install:
Screen Shot 2022-02-10 at 8 53 24 AM

**Comment by grosser** [comment]
first there was an obvious diff from not bundling all the gemfiles, maybe that already helps ...https://github.com/zendesk/stronger_parameters/pull/97

**Comment by grosser** [comment]
CI is still sad

**Comment by grosser** [comment]
might be better to get remove rails 4.2 out of the way in a separate PR ?

**Comment by skateman** [comment]
@grosser I think I have figured it out :see_no_evil: it was my bad :see_no_evil:

**Comment by skateman** [comment]
Thank you @grosser, would it be possible to push out a new gem as well?

**Comment by grosser** [comment]
2.16.0

**Pull request merged**